### DISCUSSION TITLE
How to resize the font based on the length of the text?

### DISCUSSION THREAD

**Discussion by passionetartufo** [python-pillow/Pillow#6891] [Q&A] [answered]
Hi
I have a list of inspirational quotes and I'm trying to build a python script using Pillow to print those quotes in images. Basically my goal is to create something like this ending result with Pillow https://raw.githubusercontent.com/pythad/nider/master/examples/example1/result.png

Please note that I have thousands of quotes so I need to bulk print those quotes. 
But as you can guess, quotes may be of different length. Some may be very long and may be of thousands of characters. 
I have made a script for now,  this works fine if the text is up 190 characters. But if the text is longer I start to see problems. For example, margins are getting larger for some reasons I don't know, text is being cut and goes out of the image, sometimes I see weird errors. 
Personally I don't know how to handle this situation. what is the best solution to make my quote to always fit into the image? I was thinking if maybe Pillow has an autoresize function for the font size or something like that?

**Comment by radarhere** [comment, marked as answer]
Hi. Pillow doesn't have an autoresize function, but it does have functions to determine how large text will be once it is drawn - you might be interested in [`getlength()`](https://pillow.readthedocs.io/en/stable/reference/ImageFont.html#PIL.ImageFont.ImageFont.getlength), [`getbbox()`](https://pillow.readthedocs.io/en/stable/reference/ImageFont.html#PIL.ImageFont.ImageFont.getbbox) or [`multiline_textbbox()`](https://pillow.readthedocs.io/en/stable/reference/ImageDraw.html#PIL.ImageDraw.ImageDraw.multiline_textbbox).

Here's a demonstration.
```python
from PIL import Image, ImageDraw, ImageFont
import urllib.request

req = urllib.request.urlopen("https://python-pillow.org/dist/5d1875f38a1f87b7dbc3.png")
width = 750
height = 500
im = Image.open(req).resize((width, height))
draw = ImageDraw.Draw(im, "RGBA")

quote = "Pillow is the friendly PIL fork by Alex Clark and Contributors. PIL is the Python Imaging Library by Fredrik Lundh and Contributors. As of 2019, Pillow development is supported by Tidelift. The Python Imaging Library adds image processing capabilities to your Python interpreter. This library provides extensive file format support, an efficient internal representation, and fairly powerful image processing capabilities. The core image library is designed for fast access to data stored in a few basic pixel formats. It should provide a solid foundation for a general image processing tool."
text_width = width * 0.8
text_max_height = height * 0.8

size = 36
while size > 9:
	# Insert your own font path here
	font_path = "Times New Roman.ttf"
	font = ImageFont.truetype(font_path, size, layout_engine=ImageFont.Layout.BASIC)
	lines = []
	line = ""
	for word in quote.split():
		proposed_line = line
		if line:
			proposed_line += " "
		proposed_line += word
		if font.getlength(proposed_line) <= text_width:
			line = proposed_line
		else:
			# If this word was added, the line would be too long
			# Start a new line instead
			lines.append(line)
			line = word
	if line:
		lines.append(line)
	quote = "\n".join(lines)

	x1, y1, x2, y2 = draw.multiline_textbbox((0, 0), quote, font, stroke_width=2)
	w, h = x2 - x1, y2 - y1
	if h <= text_max_height:
		break
	else:
		# The text did not fit comfortably into the image
		# Try again at a smaller font size
		size -= 1

draw.multiline_text((width / 2 - w / 2 - x1, height / 2 - h / 2 - y1), quote, font=font, align="center", stroke_width=2, stroke_fill="#000")
im.save("out.png")
```
![out](https://user-images.githubusercontent.com/3112309/212465882-d2b0e4e0-9833-4808-949b-5eb8d1e6a137.png)

**Reply by radarhere** [reply]
@giggioman00 did that answer your question?

**Reply by Tetrax-10** [reply]
@radarhere How to do this just for one line?

https://github.com/python-pillow/Pillow/discussions/7914

**Reply by nulano** [reply]
@Tetrax-10 You can just replace the while loop with something like this:

```python
size = 36
while size > 9:
	font = ImageFont.truetype("ariblk.ttf", size)
	if font.getlength(text) <= text_width:
		break
	else:
		size -= 1
```
